pkg/eventmonitor/consumers/process.go

```diff
@@ -0,0 +1,171 @@
+	}
+
+	if err := evm.AddEventConsumerHandler(pc); err != nil {
+		return nil, fmt.Errorf("cannot add event consumer handler: %w", err)
+	}
+
+	evm.RegisterEventConsumer(pc)
+
+	return pc, nil
+}
+
+// --- eventmonitor.EventConsumer interface methods
+
+// Start starts the consumer, this is a no-op for this implementation
+func (p *ProcessConsumer) Start() error {
+	return nil
+}
+
+// Stop stops the consumer, this is a no-op for this implementation
+func (p *ProcessConsumer) Stop() {
+}
+
+// ID returns the ID of the consumer
+func (p *ProcessConsumer) ID() string {
+	return p.id
+}
+
+// --- eventmonitor.EventConsumerHandler interface methods
+
+// ChanSize returns the size of the channel that the event monitor will use to send events to this handler
+func (p *ProcessConsumer) ChanSize() int {
+	return p.chanSize
+}
+
+// EventTypes returns the event types that this handler is interested in
+func (p *ProcessConsumer) EventTypes() []model.EventType {
+	return []model.EventType{model.ExecEventType, model.ExitEventType}
+}
+
+// Copy copies the event from the eventmonitor type to the one we use in this struct
+func (p *ProcessConsumer) Copy(ev *model.Event) any {
+	return &event{
+		eventType: ev.GetEventType(),
+		pid:       ev.GetProcessPid(),
+	}
+}
+
+// HandleEvent handles the event from the event monitor
+func (p *ProcessConsumer) HandleEvent(ev any) {
+	sevent, ok := ev.(*event)
+	if !ok {
+		return
+	}
+
+	switch sevent.eventType {
+	case model.ExecEventType:
+		p.execCallbacks.call(sevent.pid)
+	case model.ExitEventType:
+		p.exitCallbacks.call(sevent.pid)
+	}
+}
+
+// SubscribeExec subscribes to process exec events, and returns the function
+// that needs to be called to unsubscribe
+func (p *ProcessConsumer) SubscribeExec(callback ProcessCallback) func() {
+	return p.execCallbacks.add(callback)
+}
+
+// SubscribeExit subscribes to process exit events, and returns the function
+// that needs to be called to unsubscribe
+func (p *ProcessConsumer) SubscribeExit(callback ProcessCallback) func() {
+	return p.exitCallbacks.add(callback)
+}
+
+// callbackMap is a helper struct that holds a map of callbacks and a mutex to protect it
+type callbackMap struct {
+	// callbacks holds the set of callbacks
+	callbacks map[*ProcessCallback]struct{}
+
+	// mutex is the mutex that protects the callbacks map
+	mutex sync.RWMutex
+
+	// hasCallbacks is a flag that indicates if there are any callbacks subscribed, used
+	// to avoid locking/unlocking the mutex if there are no callbacks
+	hasCallbacks atomic.Bool
+}
+
+// add adds a callback to the callback map and returns a function that can be called to remove it
+func (c *callbackMap) add(cb ProcessCallback) func() {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.callbacks[&cb] = struct{}{}
+	c.hasCallbacks.Store(true)
+
+	return func() {
+		c.mutex.Lock()
+		defer c.mutex.Unlock()
+		delete(c.callbacks, &cb)
+		c.hasCallbacks.Store(len(c.callbacks) > 0)
```

Comment:
i am not sure this is really "thread safe". consider the following flow
1. [go routine 1] calculate the `len(c.callback)` and get 0
2. switch to [go routine 2]
3. [go routine 2] add additional callback
4.  switch back to [go routine 1]
5. update the hasCallback to `false` (while there is a callback stored)